### PR TITLE
Ma/multiple keys

### PR DESCRIPTION
**Pull request by manderson26**
@opscode/server-team 
I'll squash before merging.

[Omnibus Branch](https://github.com/opscode/opscode-omnibus/pull/659)
[Build](http://wilson.ci.chef.co/job/chef-server-12-trigger-ad_hoc/258/downstreambuildview/)

### CODE REVIEW

**Comment by mmzyk** [comment]
@doubt72 Yeah, I agree with that. I just wanted to mention some thoughts on larger topics just to ensure they are being thought about so that we don't somehow paint ourselves into the corner. I understand if the larger topics are not tackled immediately.

File: deploy/multiple_keys.sql [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+-- 
+-- This needs some benchmarking on real data sizes; this may trigger sequential scans
+-- when we do something like
+-- select id,org_id,name,type from keys_by_name where name ='wei' AND (org_id='global' OR org_id='181131e4e896ec930f5ca3f36b490c1f');; 
+-- 
+CREATE OR REPLACE view keys_by_name AS 
+SELECT id, org_id, name, authz_id, 'client' AS type, key_id, keys.public_key, key_version FROM clients INNER JOIN keys USING (id) 
+UNION ALL
+SELECT id, 'global', username AS name, authz_id, 'user' AS type, key_id, keys.public_key, key_version FROM users INNER JOIN keys USING (id);
+
```
Comment (mmzyk) [inline deploy/multiple_keys.sql]: I was thinking of UUID conflict, but I wasn't 100% certain that the ids in this case were UUIDs. Given that my worry of id conflicts would be overblown, as you said.

File: deploy/multiple_keys.sql [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+--
+
+BEGIN;
+
+CREATE TABLE keys(
+   id char(32) NOT NULL,
+   key_id text NOT NULL,
+   public_key text NOT NULL,
+   key_version integer NOT NULL,
+   created_at TIMESTAMP WITHOUT TIME ZONE NOT NULL,
```
Comment (mmzyk) [inline deploy/multiple_keys.sql]: So based on James' linked article, it sounds like if with went with timestamptz we'd also want to make sure that we set the default timezone to UTC, if we've not already done so. (The author of that article happens to be the author of sqitch, fyi).

I could also see an argument for leaving as is for consistency with the rest of the chef server schema, if other fields are not storing the time zone.

File: deploy/multiple_keys_migration.sql [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+--
+-- Initial migration
+-- The 'WHERE NOT EXISTS' clauses should prevent inserts when we already have a key.
+--
+
+LOCK TABLE keys IN SHARE ROW EXCLUSIVE MODE;
+  -- add clients
+  INSERT INTO keys
+    (id, key_id, public_key, key_version, created_at,  expires_at)
+	SELECT id, 'default', public_key, pubkey_version, updated_at, 'infinity'::timestamp FROM clients
```
Comment (marcparadise) [inline deploy/multiple_keys_migration.sql]: Let's talk about pubkey_version and if we really need it. I'd like to move away from the branching on certificate(0?)/pub key (1?) and just force pub key.  Would this make sense as the right time to do that?

**Pull request merged**